### PR TITLE
Less aggressively poison sources on builds

### LINKED ISSUE #5257: Regression: Cargo nightly updates registry on every build with "path" crate
When one of the dependencies is local, specified using "path", Cargo updates the registry on every build.

I think this was supposed to be fixed, but is present with nightly Rust (but not stable).

### PR DESCRIPTION
**Pull request by alexcrichton**
Discovered in #5257 the changes in #5215 were slightly too aggressively
poisoning sources to require updates, thinking that a manifest changed when it
actually hadn't.

Non-workspace-member path dependencies with optional/dev-dependencies
don't show up in the lock file, so the previous logic would recognize this and
think that the dependency missing from the lock file was just added and would
require a registry update.

The fix in this commit effectively just skips all of these dependencies in
non-workspace members. This means that this will be slightly buggy if an
optional dependency that's activated is added, but that's hopefully something we
can tackle later.

Closes #5257

### CODE REVIEW

**Comment by alexcrichton** [comment]
I think we'll want to backport this to beta as well

File: tests/testsuite/freshness.rs [hunk 1]
```diff
@@ -1007,3 +1008,47 @@ fn no_rebuild_when_rename_dir() {
...
+        .file(
+            "foo/Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.1.1"
+                authors = []
+
+                #[dependencies]
+                #registry2 = { version = "*", optional = true }
```
Comment (matklad) [inline tests/testsuite/freshness.rs]: Why is this commented out?

File: tests/testsuite/freshness.rs [hunk 2]
```diff
@@ -1007,3 +1008,47 @@ fn no_rebuild_when_rename_dir() {
...
+                [dev-dependencies]
+                registry2 = "*"
+            "#,
+        )
```
Comment (matklad) [inline tests/testsuite/freshness.rs]: No need for raw literal

File: tests/testsuite/freshness.rs (hunk 1 above)
Comment (alexcrichton) [reply tests/testsuite/freshness.rs]: Oh I left that in by accident, either version of this triggers a bug and I was confirming that they both do

**Comment by alexcrichton** [comment]
Updated!

File: tests/testsuite/freshness.rs [hunk 3]
```diff
@@ -1007,3 +1008,45 @@ fn no_rebuild_when_rename_dir() {
...
+        .file(
+            "foo/Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.1.1"
+                authors = []
+
+                [dev-dependencies]
+                registry2 = "*"
```
Comment (matklad) [inline tests/testsuite/freshness.rs]: So this checks only non-transitive case? Let’s add a separate test for optional dep as well?

**Comment by matklad** [comment]
Will do the backports!

**Pull request merged**